### ISSUE TITLE
VayraPersonBountyManager causes ConcurrentModificationException

### ISSUE THREAD

**Issue by FERNANDOHYDE** [RkShaRkz/Starsector-IntelliJ-Template#30]
I get the following error starting a campaign on linux 

`143302 [Thread-4] ERROR com.fs.starfarer.combat.CombatMain  - java.util.ConcurrentModificationException
java.util.ConcurrentModificationException
        at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:911)
        at java.util.ArrayList$Itr.next(ArrayList.java:861)
        at data.scripts.campaign.intel.VayraPersonBountyManager.checkForFuckedUpParticipants(VayraPersonBountyManager.java:76)
        at data.scripts.campaign.intel.VayraPersonBountyManager.advance(VayraPersonBountyManager.java:268)
        at data.scripts.VayraMergedModPlugin.onNewGameAfterTimePass(VayraMergedModPlugin.java:343)
        at com.fs.starfarer.campaign.save.CampaignGameManager.o00000(Unknown Source)
        at com.fs.starfarer.title.TitleScreenState.dialogDismissed(Unknown Source)
        at com.fs.starfarer.ui.O.dismiss(Unknown Source)
        at com.fs.starfarer.ui.impl.OO0OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO.dismiss(Unknown Source)
        at com.fs.starfarer.campaign.save.J.actionPerformed(Unknown Source)
        at com.fs.starfarer.ui.n.buttonPressed(Unknown Source)
        at com.fs.starfarer.ui.int.Ò00000(Unknown Source)
        at com.fs.starfarer.ui.int.processInput(Unknown Source)
        at com.fs.starfarer.ui.o0Oo.o00000(Unknown Source)
        at com.fs.starfarer.BaseGameState.traverse(Unknown Source)
        at com.fs.state.AppDriver.begin(Unknown Source)
        at com.fs.starfarer.combat.CombatMain.main(Unknown Source)
        at com.fs.starfarer.StarfarerLauncher.o00000(Unknown Source)
        at com.fs.starfarer.StarfarerLauncher$1.run(Unknown Source)
        at java.lang.Thread.run(Thread.java:750)
`

**Comment by RkShaRkz** [comment]
This was fixed by #21 - closing. Though perhaps I should release a new version...

**Comment by RkShaRkz** [comment]
But you know what, let me push out a different fix for this issue just in case.

**Comment by RkShaRkz** [comment]
deployed a different fix in [8bfd2b2](https://github.com/RkShaRkz/Starsector-IntelliJ-Template/commit/8bfd2b2ae349f1d6c96305d25b75a870c2880fd6)

**Comment by RkShaRkz** [comment]
@FERNANDOHYDE if you manage to repro on 3.3.9 please ping me to reopen this and take a look at it more deeper. From what i'm seeing right now:
```
    private void checkForFuckedUpParticipants() {
        checkedForFuckedUpParticipants = true;
        for (Iterator<String> iter = getSharedData().getParticipatingFactions().iterator(); iter.hasNext(); ) {    // line 76
            String factionId = iter.next();
            if (Global.getSector().getFaction(factionId) == null) {
                Global.getSector().getCampaignUI().addMessage(factionId + " is an invalid bounty participant, please yell at its mod author", Color.red);
                Global.getSector().getCampaignUI().addMessage(factionId + " bounties will not function until they fix it", Color.red);
                Global.getSector().getCampaignUI().addMessage("i could have crashed the game right here, i SHOULD have crashed the game right here", Color.red);
                Global.getSector().getCampaignUI().addMessage("god knows you'd fuckin' deserve it", Color.red);
                iter.remove();
            }
        }
    }
```
nothing in here should trigger the `ConcurrentModificationException` unless something else is adding to / removing from the list.
If it reproes i'll try something different, which, again won't guarantee that the ConcurrentModificationException will be gone at all because it shouldn't be happening in this code to begin with.

But if some other mod/thread is modifying this same list (e.g. MagicLib's bounty manager which has caused problems 0 times for me yet), there is no way for me to introduce synchronization on the underlying list or change it from an ArrayList to CopyOnWriteList or some other synchronized data structure, hence why I say - if you repro on new version, the other fix I have in mind can't guarantee not causing it either.

**Comment by RkShaRkz** [comment]
sorry, upon a better inspection i see the list is accessed from multiple files and since i don't know if they're ran in single-threaded or multi-threaded fashion, lemme make all accesses to this bullshit synchronized.

**Comment by RkShaRkz** [comment]
Alternative fix pending in #37 - but just so I don't end up merging it in too eagerly, here's a hotfix candidate you should try out first, and report back if it fixes the issue or not.

https://ufile.io/kbu9jjo3

cc @FERNANDOHYDE

**Comment by RkShaRkz** [comment]
@FERNANDOHYDE any update on the hotfix? did it resolve the issue?

**Comment by RkShaRkz** [comment]
Since #37 is merged in, and this issue is duplicate of issues #18 #19 #20   and was first taken care of with fix from PR #21 and again with #37 - I'm closing this until it resurfaces again on 3.3.11
